File: app/src/main/java/com/example/swob_deku/Models/GatewayClients/GatewayClientViewModel.java
package com.example.swob_deku.Models.GatewayClients;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

public class GatewayClientViewModel extends ViewModel {

    private MutableLiveData<List<GatewayClient>> gatewayClientList;
    GatewayClientDAO gatewayClientDAO;

    public MutableLiveData<List<GatewayClient>> getGatewayClientList(Context context, GatewayClientDAO gatewayClientDAO) {
        if(gatewayClientList == null) {
            this.gatewayClientDAO = gatewayClientDAO;
            gatewayClientList = new MutableLiveData<>();
            loadGatewayClients(context);
        }
        return gatewayClientList;
    }

    public void refresh(Context context) {
        loadGatewayClients(context);
    }

    private void loadGatewayClients(Context context) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                List<GatewayClient> gatewayClients = gatewayClientDAO.getAll();
                Log.d(getClass().getName(), "Number of items: " + gatewayClients.size());

                if(gatewayClients != null)
                    for(GatewayClient gatewayClient : gatewayClients)
                        gatewayClient.setConnectionStatus(
                                GatewayClientHandler.getConnectionStatus(context,
                                        String.valueOf(gatewayClient.getId())));

                gatewayClientList.postValue(gatewayClients);
            }
        }).start();
    }
}

File: app/src/main/java/com/afkanerd/deku/QueueListener/GatewayClients/GatewayClientViewModel.kt
package com.afkanerd.deku.QueueListener.GatewayClients

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.afkanerd.deku.Datastore
import com.afkanerd.deku.Modules.ThreadingPoolExecutor

class GatewayClientViewModel : ViewModel() {
    private var gatewayClientList: LiveData<List<GatewayClient>> = MutableLiveData()

    private lateinit var datastore: Datastore

    fun getGatewayClientList(context: Context): LiveData<List<GatewayClient>> {
        datastore = Datastore.getDatastore(context)
        if(gatewayClientList.value.isNullOrEmpty()) {
            gatewayClientList = loadGatewayClients()
        }
        return gatewayClientList
    }

    private fun loadGatewayClients() : LiveData<List<GatewayClient>> {
        return datastore.gatewayClientDAO().fetch()
    }

    fun update(gatewayClient: GatewayClient) {
        ThreadingPoolExecutor.executorService.execute {
            datastore.gatewayClientDAO().update(gatewayClient)
        }
    }

//    private fun normalizeGatewayClients(gatewayClients: List<GatewayClient>): List<GatewayClient> {
//        val filteredGatewayClients: MutableList<GatewayClient> = ArrayList()
//        for (gatewayClient in gatewayClients) {
//            var contained = false
//            for (gatewayClient1 in filteredGatewayClients) {
//                if (gatewayClient1.same(gatewayClient)) {
//                    contained = true
//                    break
//                }
//            }
//            if (!contained) {
//                filteredGatewayClients.add(gatewayClient)
//            }
//        }
//
//        return filteredGatewayClients
//    }
}
